app/scripts/controllers/permissions/index.js

```diff
@@ -510,6 +507,100 @@ export class PermissionsController {
     })
   }
 
+  /**
+   * Removes all known domains and their related permissions.
+   */
+  clearPermissions () {
+    this.permissions.clearDomains()
+    this.notifyAllDomains({
+      method: NOTIFICATION_NAMES.accountsChanged,
+      result: [],
+    })
+  }
+
+  /**
+   * Stores domain metadata for the given origin (domain).
+   * Deletes metadata for domains without permissions in a FIFO manner.
+   * Metadata is only deleted for domains with permissions on boot in order to
+   * prevent a degraded user experience, as metadata currently can't be
+   * requested on-demand.
+   *
+   * @param {string} origin - The origin whose domain metadata to store.
+   * @param {Object} metadata - The domain's metadata that will be stored.
+   */
+  addDomainMetadata (origin, metadata) {
+
+    const oldMetadataState = this.store.getState()[METADATA_STORE_KEY]
+    const newMetadataState = { ...oldMetadataState }
+
+    // delete pending metadata origin from queue, and delete its metadata if
+    // it doesn't have any permissions
+    if (this._pendingSiteMetadata.size >= METADATA_CACHE_MAX_SIZE) {
```

Comment:
This should first check to see whether the metadata for this site is already stored. If it is, all of this can be skipped.

The way it's implemented now, we might be deleting and then re-adding the same thing each page load (if it happens to be the oldest thing in the cache anyway).

Reply:
Well, deleting and re-adding origins from the `_pendingSiteMetadata` set is a feature, not a bug, because we want the insertion order to be updated.

We could end up deleting and re-adding the `siteMetadata`, too, but that'll have to get overwritten anyway (it may have changed), and `delete` is cheap.

Reply:
Hmm. OK. Well that's not ideal from a performance standpoint, but I guess it's negligible next to our other performance issues! I suppose this is OK for now.